File: src/services/notes.ts
/**
 * Per-image free-text annotations (Phase 11). Backed by the
 * `notes` column on the images table; IPC commands are
 * `get_image_notes` / `set_image_notes`.
 *
 * Empty string at the user-facing level means "no annotation"; the
 * backend collapses NULL and "" into the same Option<String>::None.
 */
import { invoke } from "@tauri-apps/api/core";

export async function getImageNotes(imageId: number): Promise<string> {
  try {
    return await invoke<string>("get_image_notes", { imageId });
  } catch (error) {
    throw new Error(`Failed to get notes: ${error}`);
  }
}

export async function setImageNotes(
  imageId: number,
  notes: string,
): Promise<void> {
  try {
    await invoke("set_image_notes", { imageId, notes });
  } catch (error) {
    throw new Error(`Failed to save notes: ${error}`);
  }
}

File: src-tauri/src/commands/notes.rs
use tauri::State;

use crate::commands::ApiError;
use crate::db::ImageDatabase;

/// Read the free-text annotation for an image. Returns "" if there
/// is no annotation set (the column is either NULL or "" — we treat
/// both as "no annotation" at the user-facing level).
#[tauri::command]
pub fn get_image_notes(
    db: State<'_, ImageDatabase>,
    image_id: i64,
) -> Result<String, ApiError> {
    Ok(db.get_image_notes(image_id)?.unwrap_or_default())
}

/// Write an annotation for an image. Empty / whitespace-only string
/// clears the field.
#[tauri::command]
pub fn set_image_notes(
    db: State<'_, ImageDatabase>,
    image_id: i64,
    notes: String,
) -> Result<(), ApiError> {
    Ok(db.set_image_notes(image_id, &notes)?)
}
